src/util/reference_processor.rs

```diff
@@ -129,207 +162,392 @@ pub enum Semantics {
+                    new_set.insert(new_reference);
                 }
             }
-            trace!(
-                "{:?} references: {} -> {}",
+            debug!(
+                "{:?} reference table from {} to {} ({} enqueued)",
                 self.semantics,
-                references.len(),
-                to_index
+                sync.references.len(),
+                new_set.len(),
+                enqueued_references.len()
             );
-            sync.nursery_index = to_index;
-            references.truncate(to_index);
+            sync.references = new_set;
+            sync.enqueued_references = enqueued_references;
         }
 
-        /* flush out any remset entries generated during the above activities */
-
-        // FIXME: We are calling mutator() for a worker thread
-        panic!("We are calling mutator() for a worker tls. We need to fix this.");
-        // unsafe { VM::VMActivePlan::mutator(tls)) }.flush_remembered_sets();
-        // if TRACE {
-        //     trace!("Ending ReferenceProcessor.scan({:?})", self.semantics);
-        // }
+        debug!("Ending ReferenceProcessor.scan({:?})", self.semantics);
     }
 
-    /**
-     * This method deals only with soft references. It retains the referent
-     * if the reference is definitely reachable.
-     * @param reference the address of the reference. This may or may not
-     * be the address of a heap object, depending on the VM.
-     * @param trace the thread local trace element.
-     */
-    fn retain_referent<VM: VMBinding, T: TraceLocal>(
-        &self,
-        trace: &mut T,
-        reference: ObjectReference,
-    ) {
+    /// This method deals only with soft references. It retains the referent
+    /// if the reference is definitely reachable.
+    fn retain_referent<E: ProcessEdgesWork>(&self, trace: &mut E, reference: ObjectReference) {
         debug_assert!(!reference.is_null());
         debug_assert!(self.semantics == Semantics::SOFT);
 
-        if TRACE_DETAIL {
-            trace!("Processing reference: {:?}", reference);
-        }
+        trace!("Processing reference: {:?}", reference);
 
         if !reference.is_live() {
-            /*
-             * Reference is currently unreachable but may get reachable by the
-             * following trace. We postpone the decision.
-             */
+            // Reference is currently unreachable but may get reachable by the
+            // following trace. We postpone the decision.
             return;
         }
 
-        /*
-         * Reference is definitely reachable.  Retain the referent.
-         */
-        let referent = VM::VMReferenceGlue::get_referent(reference);
+        // Reference is definitely reachable.  Retain the referent.
+        let referent = <E::VM as VMBinding>::VMReferenceGlue::get_referent(reference);
         if !referent.is_null() {
-            trace.retain_referent(referent);
+            Self::keep_referent_alive(trace, referent);
```

Comment:
`keep_referent_alive` calls `e.trace_object`, and it may move the object, and return the forwarded reference.  But we are not calling `VMReferenceGlue::set_referent` here.  Currently the two call sites of `set_referent` are:

1.  in `forward` for MarkCompact.
2.  in `process_reference`.  But `scan` calls `process_reference` when `retain` is `false`.  But this line is executed when `retain` is `true`.

So if we use the `SemiSpace` plan, and a `SoftReference` holds the only pointer to another object, and this is the first time this edge is traced, will this `SoftReference` hold a dangling pointer after that? 

```suggestion
            let new_referent = Self::keep_referent_alive(trace, referent);
            <E::VM as VMBinding>::VMReferenceGlue::set_referent(reference, new_referent);
```

Reply:
That is not necessary. However, the code is a bit confusing. What actually happens is this:
1. In `SoftRefClosure`:
    * if emergency collection { retain soft refs (no update) }
3. In `WeakRefClosure`:
    * scan soft refs (with update)
    * scan weak refs (with update)

I will change this to:
1. In `SoftRefClosure`:
    * if emergency collection { retain soft refs (no update) }
    * scan soft refs (with update)
4. In `WeakRefClosure`:
    * scan weak refs (with update)